public/app/features/dashboard/components/FolderPicker/FolderPicker.tsx

```diff
@@ -0,0 +1,266 @@
+import React, { PureComponent } from 'react';
+import { Button, Input, AsyncSelect } from '@grafana/ui';
+import { AppEvents, SelectableValue } from '@grafana/data';
+import { getBackendSrv } from 'app/core/services/backend_srv';
+import { appEvents, contextSrv } from 'app/core/core';
+import validationSrv from 'app/features/manage-dashboards/services/ValidationSrv';
+
+export interface Props {
+  onChange: ($folder: { id: number; title: string }) => void;
+
+  labelClass?: string;
+  rootName?: string;
+  enableCreateNew?: boolean;
+  enableReset?: boolean;
+  dashboardId?: any;
+  initialTitle?: string;
+  initialFolderId?: number;
+
+  enterFolderCreation?: () => void;
+  exitFolderCreation?: () => void;
+}
+
+interface State {
+  folder: SelectableValue<number>;
+  newFolderName: string;
+  createNewFolder: boolean;
+  validationError: string;
+  hasValidationError: boolean;
+  newFolderNameTouched: boolean;
+  isLoading: boolean;
+}
+
+export class FolderPicker extends PureComponent<Props, State> {
+  static defaultProps = {
+    rootName: 'General',
+    enableCreateNew: false,
+    enableReset: false,
+    initialTitle: '',
+  };
+
+  state: State = {
+    folder: {},
+    newFolderName: '',
+    createNewFolder: false,
+    validationError: '',
+    hasValidationError: false,
+    newFolderNameTouched: false,
+    isLoading: true,
+  };
+
+  componentDidMount = async () => {
+    await this.loadInitialValue();
```

Comment:
Maybe consider moving all async operations to Redux action and move state to Redux state?